server/src/modules/users/test/users.service.test.ts

```diff
@@ -0,0 +1,575 @@
+import { UsersService } from '../users.service';
+import { getModelToken } from '@nestjs/mongoose';
+import { UserDocument } from '../../../schemas/user.schema';
+import { HttpException, HttpStatus } from '@nestjs/common';
+import { CreateUserDto, UpdateUserDto } from '../users.dto';
+import { UserAuthProvider } from 'src/enums/user.auth.provider';
+import { UserRole } from 'src/enums/user.role';
+import { SlideObjectDocument } from 'src/schemas/slide-object.schema';
+
+class UserModelMock {
+  find = jest.fn();
+  findOne = jest.fn();
+  findById = jest.fn();
+  findByIdAndUpdate = jest.fn();
+  findByIdAndDelete = jest.fn();
+  create = jest.fn();
+  save = jest.fn();
+}
+
+describe('UsersService', () => {
+  // Setup
+
+  let userService: UsersService;
+  const userModelMock = new UserModelMock();
+
+  beforeAll(async () => {
+    const module: TestingModule = await Test.createTestingModule({
+      providers: [
+        UsersService,
+        {
+          provide: getModelToken('User'),
+          useValue: userModelMock,
+        },
+      ],
+    }).compile();
+
+    userService = module.get<UsersService>(UsersService);
+  });
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+
+  // Data
+
+  const userId1 = '1';
+  const userEmail1 = 'test1@example.com';
+  const userPassword1 = 'P4$$wOrD1234';
+  const userRole1 = UserRole.USER;
+  const userAuthProvider1 = UserAuthProvider.INTERNAL_AND_EXTERNAL;
+
+  const userDocument1 = {
+    _id: userId1,
+    email: userEmail1,
+    password: userPassword1,
+    userRole: userRole1,
+    userAuthProvider: userAuthProvider1,
+  } as UserDocument;
+
+  const createUserDto1 = {
+    email: userEmail1,
+    password: userPassword1,
+  } as CreateUserDto;
+
+  const updatedUserEmail1 = 'updated@mail.com';
+  const updatedUserPassword1 = 'updatedP4$$woRD';
+
+  const updateUserDto1 = {
+    email: updatedUserEmail1,
+    password: updatedUserPassword1,
+  } as UpdateUserDto;
+
+  const updatedUserId1 = '1';
+  const updatedUserRole1 = UserRole.ADMIN;
+  const updatedUserAuthProvider1 = UserAuthProvider.GOOGLE;
+
+  const updatedUserDocument1: UserDocument = {
+    _id: updatedUserId1,
+    email: updatedUserEmail1,
+    password: updatedUserPassword1,
+    userRole: updatedUserRole1,
+    userAuthProvider: updatedUserAuthProvider1,
+  } as UserDocument;
+
+  const userId2 = '2';
+  const userEmail2 = 'test2@mail.com';
+  const userPassword2 = 'anotherP4$$woRD';
+  const userRole2 = UserRole.ADMIN;
+  const userAuthProvider2 = UserAuthProvider.GOOGLE;
+
+  const userDocument2 = {
+    _id: userId2,
+    email: userEmail2,
+    password: userPassword2,
+    userRole: userRole2,
+    userAuthProvider: userAuthProvider2,
+  } as UserDocument;
+
+  const userDocumentsArray = [userDocument1, userDocument2];
+
+  const boardId1 = '222';
+
+  const slideObject1 = { height: 20.0, width: 22.2 } as SlideObjectDocument;
+
+  const slideObjectId1 = '2424';
+
+  // Tests
+
+  describe('findUsers', () => {
```

Comment:
I thought we should test methods that are using helpers, for example
getUsers, getUserById, createUser, updateUser, deleteUser, etc.

Reply:
I dont see point testing database operations then

Reply:
Well, we are testing business logic behind our service, not database operations. Database operations are mocked.